### DISCUSSION TITLE
Command for adding a space to the beginning of each line of the Clipboard

### DISCUSSION THREAD

**Discussion by francwalter** [hluk/CopyQ#2571] [Q&A] [answered]
Halo
I try to create a very simple command to add just a space to the beginning of each line of my clipboard.
I tried this:

```
[Command]
Name=Add a space
Command="
    copyq:
    var text = str(clipboard())
    if (text) {
        text = text.replace(new RegExp("^", "g"), " ")
        copy(text)
        paste()
    }"
IsGlobalShortcut=true
Icon=\xf105
GlobalShortcut=meta+alt+shift+space
```

but it adds a space only to the beginning of the first line of my clipboard's text, not to the following lines.
I tried also:

`text = text.replace(/^/gm, ' ');`

but still not right. I even dont know where to find the RegEx documentation, which one CopyQ uses etc. (EDIT: ECMA Script, as I think now).

Later I would like to add a command to **delete** the first char. at the beginning of each line, but only if I first get that space adding right...

Thanks for hints.
frank

EDIT: but this works on the [Try it from MDN Webdocs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp)

```
const text = new String("First line,\nsecond line\nthird line");
console.log(text.replace(/^/gm, "  "));
```

I thought that the language of commands is ECMA Script like JS?

**Comment by francwalter** [comment]
OK then, I found it myself by looking into other command scripts, especially the Find & Replace did my trick :)
Now I have two handy commands, one for adding a space each line, nice for indents in coding, where it is not supported by the editor, and another one to delete the first character of each line, sometimes useful, when I have e.g. a list of files with full path and only need the file names, e.g.
So here my commands. First the "Add as space at line beginning":

```
[Command]
Name=Add a space at line beginning
Command="
    copyq:
    // 2023-12-13: von Search & Replace entlehnt, siehe:
    // https://github.com/hluk/CopyQ/discussions/2571
    // https://copyq.readthedocs.io/en/latest/scripting-api.html
    var sel = ItemSelection();
    sel = sel.current();

    for (var index = 0; index < sel.length; ++index) {
        var item = sel.itemAtIndex(index);
        const text = str(item[mimeText]);
        const newText = text.replace(/^/gm, ' ');
        if (text != newText) {
            item[mimeText] = newText;
            delete item[mimeHtml];
            sel.setItemAtIndex(index, item);
        }
    }"
InMenu=true
IsGlobalShortcut=true
Icon=\xf105
Shortcut=meta+alt+space
```

And the "Delete first char of line"

```
[Command]
Name=Delete first char of line
Command="
    copyq:
    // 2023-12-13: von Search & Replace entlehnt, siehe:
    // https://github.com/hluk/CopyQ/discussions/2571
    // https://copyq.readthedocs.io/en/latest/scripting-api.html
    var sel = ItemSelection();
    sel = sel.current();

    for (var index = 0; index < sel.length; ++index) {
        var item = sel.itemAtIndex(index);
        const text = str(item[mimeText]);
        const newText = text.replace(/^./gm, '');
        if (text != newText) {
            item[mimeText] = newText;
            delete item[mimeHtml];
            sel.setItemAtIndex(index, item);
        }
    }"
InMenu=true
IsGlobalShortcut=true
Icon=\xf104
Shortcut=meta+alt+-
```

Both are just directly changing the selected (or newest) clipboard list entry. So no backup of the original clip, which is not neccessary for me anyway.
Thanks.frank

**EDIT 2023-12-19:** here the code to remove commented code lines (only // and #). Sometimes I have e.g. config files with huge ammounts of comments (by dist or so) and only very few real config lines. To get just these "real" code lines, I can clean them with that command:

```
copyq:
// Delete comments lines in code (// or #)
// 2023-12-19: adapted from Search & Replace
var sel = ItemSelection();
sel = sel.current();

for (var index = 0; index < sel.length; ++index) {
    var item = sel.itemAtIndex(index);
    const text = str(item[mimeText]);
    const newText = text.replace(/^\s*(\/\/|#)+[^\n]*\n/gm, '');
    if (text != newText) {
        item[mimeText] = newText;
        delete item[mimeHtml];
        sel.setItemAtIndex(index, item);
    }
}
```

**Reply by francwalter** [reply]
And another one: 

With `const newText = text.replace(/^\s/gm, '');` instead of `const newText = text.replace(/^./gm, '');` only the beginning space of each line of the selected clip is deleted.

**Comment by francwalter** [comment]
> ```
> [Command]
> Name=Add a space at line beginning
> Command="
>     copyq:
>     // 2023-12-13: von Search & Replace entlehnt, siehe:
>     // https://github.com/hluk/CopyQ/discussions/2571
>     // https://copyq.readthedocs.io/en/latest/scripting-api.html
>     var sel = ItemSelection();
>     sel = sel.current();
> 
>     for (var index = 0; index < sel.length; ++index) {
>         var item = sel.itemAtIndex(index);
>         const text = str(item[mimeText]);
>         const newText = text.replace(/^/gm, ' ');
>         if (text != newText) {
>             item[mimeText] = newText;
>             delete item[mimeHtml];
>             sel.setItemAtIndex(index, item);
>         }
>     }"
> InMenu=true
> IsGlobalShortcut=true
> Icon=\xf105
> Shortcut=meta+alt+space
> ```

There is something weird, not working as expected in this command: it adds one space to the very first line, but two spaces to each of the following lines. I dont understand that.

**Comment by francwalter** [comment]
Is anybody here in this "Discussions" who could tell me in this very simple script what I am doing wrong?

Why is the `text.replace(/^/gm, ' ')` doing **two** replacements of spaces in the second and further lines but only one in the first line?
Is this a bug?

I tried hours now to "debug" it, but as even the console.log does not work, I dont find it 😢 

This is working in [MDN's Try it:](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace?retiredLocale=de):
```
const text = "AAAAA\nBBBBB\nCCCCC";
console.log(text.replace(/^/gm, ' '));
```
but not in the ECMA version of CopyQ 😢

**Reply by hluk** [reply]
It could be a bug in the ECMA standard implementation in Qt framework but I'm not sure how closely the multiline with caret matching is defined.

Possible workaround: `text.replace(/^|\n/gm, ' $&')`

**Comment by francwalter** [comment]
By the anchor of Sabords! It works!
Thanks a lot, Lukas!

**Comment by francwalter** [comment]
But moment, this regexp: `text.replace(/^|\n/gm, ' $&')` does a search for either line's beginning (^) or linebreak (\n) and replaces it with a space **and** the finding (either ^ or \n). No worries for the "^" but the "\n" is at the end! When it replaces it, there is a space at each end of the line! This could be mostly unimportant, but I am sure, I will have then once a replacement, where it is absolutely not useful to have (unnoticed) spaces at the end. I am using this replacement for code indentation, so spaces at the end could really be an issue.

I can make it visible, if I change my regexp like this: `text.replace(/^|\n/gm, '#$&')` then I have e.g. from that before the command:
```
AAAAA
BBBBB
CCCCC
DDDDD
```
after the command that result:
```
#AAAAA#
#BBBBB#
#CCCCC#
#DDDDD
```

**Reply by hluk** [reply, marked as answer]
Yeah, there are spaces also on empty lines after the replacements.

Maybe it is best to avoid regular expressions in this case and use something like:

```js
text.split("\n").map((line) => line ? ` ${line}` : line).join("\n")
```

**Reply by francwalter** [reply]
Now it is perfectly working, thanks ❤️
